jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/CachedQueryFields.java

```diff
@@ -0,0 +1,62 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.internal;
+
+import java.nio.charset.Charset;
+import java.util.Objects;
+
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.util.Fields;
+
+public class CachedQueryFields extends Fields
+{
+    public static final String KEY = "_oejsi_CQF";
+    private final String _query;
+    private final Charset _charset;
+
+    public static CachedQueryFields getCached(Request request, String query, Charset charset)
+    {
+        // Cached fields can be stored directly in a ChannelRequest, else use an attribute
+        CachedQueryFields cached = request instanceof HttpChannelState.ChannelRequest channelRequest
+            ? channelRequest.getCachedQueryFields()
+            : request.getAttribute(CachedQueryFields.class.getName()) instanceof CachedQueryFields c ? c : null;
+
+        return cached != null && Objects.equals(query, cached._query) && Objects.equals(charset, cached._charset) ? cached : null;
```

Comment:
Since a check on the query string and the charset is performed, I think it would be worth adding a unit test for this method.

Reply:
I have added a test.